Remove E_STRICT error level and deprecate E_STRICT constant

mvorisek opened the pull request:
`E_STRICT` error level is no longer used in the php-src and is/was already documented in the code [1] as to be removed.

`E_STRICT` error level has no use since PHP 7.4 [2] with comment in the last PR description:

> The constant still exists in the Zend engine which we may want to remove in PHP 8.

[1] https://github.com/php/php-src/blob/php-8.3.0/run-tests.php#L1647
[2] https://github.com/php/php-src/pull/4401


RFC: https://wiki.php.net/rfc/deprecations_php_8_4#remove_e_strict_error_level_and_deprecate_e_strict_constant

staabm commented:
Does this change yield any value or just create a lot of work to cleanup existing userland code without actual value?

https://github.com/search?q=%22E_STRICT%22+language%3APHP&type=code

staabm commented:
Before deleting it here IMO we should check whether static analysis tools would want to emit a error for the constant when used as a first step forward

https://phpstan.org/r/5f512016-9f2a-4387-9486-5330853e11cb

staabm commented:
Maybe we can even mark the constant as `@deprecated` in the stubs. This might already be enough to male phpstan emit error for its usages (I did not chrck whether thats true).

mvorisek commented:
I updated this PR to be BC friendly, ie. removed all E_STRICT error level related code as the error level is no longer used since PHP 7.4.

I kept the E_STRICT constant /wo even deprecating it. It might be deprecated in separate PR and such PR will probably require a RFC.

jorgsowa commented:
Deprecation of the constant may be added to the collective RFC of deprecations: https://wiki.php.net/rfc/deprecations_php_8_4

Girgias commented:
Adding it to the deprecation RFC makes sense, because in theory someone could use the constant with ``trigger_error()``

mvorisek commented:
This PR should not require RFC as it does not remove the constatnt. It removes the internal E_STRICT error level that is no longer used.

TimWolla commented:
This PR changes the user-visible value of a built-in constant, which might be considered a breaking change. This PR is probably on the line of needing an RFC and in those cases it's easier to create a succinct RFC than to argue why it would not need one. Especially if at least one contributor feels that it needs one.

mvorisek commented:
Can you or someone more involved add it to the https://wiki.php.net/rfc/deprecations_php_8_4? Thank you.

jorgsowa commented:
I have added to the list. Let me know if something is wrong.

https://wiki.php.net/rfc/deprecations_php_8_4#constant_e_strict

andypost commented:
Looks like voting finished!

iluuu1994 commented:
It looks like this is missing the actual deprecation of the constant.

Girgias commented on Zend/zend_errors.h:
```diff
@@ -39,7 +39,7 @@
 /* Indicates that this usually fatal error should not result in a bailout */
```
Add a TODO Comment about the `#define E_STRICT` to remove this definitely in PHP 9?

Girgias reviewed (changes requested):
As pointed out by @iluuu1994 the userland `E_STRICT` constant is not actually deprecated.

jrfnl commented:
Just checking as there seems to be no progress on this PR - will this make it into PHP 8.4 ? Or is this going to be delayed to PHP 8.5 ?

iluuu1994 commented:
@jrfnl Thanks for the reminder. I'll have a look.

iluuu1994 commented:
@mvorisek Ah, I see you just pushed the changes. Good timing :slightly_smiling_face: Thanks!

andypost commented:
Using 8.4.0 RC2 I see a lot of warnings in `pearcmd.php` 

```
Deprecated: Constant E_STRICT is deprecated in /usr/local/lib/php/pearcmd.php on line 441
Deprecated: Constant E_STRICT is deprecated in /usr/local/lib/php/pearcmd.php on line 441
Deprecated: Constant E_STRICT is deprecated in /usr/local/lib/php/pearcmd.php on line 441
```

is there pear repo to fix it?

jyxjjj commented:
This can be easily bypassed via change it into 2048. I am here because there's a very well known repository negatively fixes their error. I think it can be detected inside error_reporting function. Can you fix this bypass error? May I create a new issue?
